qcfractal/storage_sockets/models/collections_models.py

```diff
@@ -0,0 +1,236 @@
+import datetime
+from sqlalchemy import (Column, Integer, String, ForeignKey, JSON,
+                        ARRAY, Index)
+from sqlalchemy.orm import relationship, column_property
+# from sqlalchemy import select, func, tuple_, text, cast
+from qcfractal.storage_sockets.models import Base
+# from sqlalchemy.dialects.postgresql import array_agg
+from sqlalchemy.dialects import postgresql
+from sqlalchemy.sql.functions import GenericFunction
+from sqlalchemy.ext.hybrid import hybrid_property
+
+
+# class json_agg(GenericFunction):
+#     type = postgresql.JSON
+
+class json_build_object(GenericFunction):
+    type = postgresql.JSON
+
+
+class CollectionORM(Base):
+    """
+        A base collection class of precomuted workflows such as datasets, ..
+
+        This is a dynamic document, so it will accept any number of
+        extra fields (expandable and uncontrolled schema)
+    """
+
+    __tablename__ = "collection"
+
+    id = Column(Integer, primary_key=True)
+    collection_type = Column(String)  # for inheritance
+
+    collection = Column(String(100), nullable=False)
+    lname = Column(String(100), nullable=False)
+    name = Column(String(100), nullable=False)
+
+    tags = Column(JSON)
+    tagline = Column(String)
+
+    provenance = Column(JSON)
+
+    extra = Column(JSON)  # extra data related to specific collection type
+
+    def update_relations(self, **kwarg):
+        pass
+
+
+    __table_args__ = (
+        Index('ix_collection_lname', "collection", "lname", unique=True),
+        Index('ix_collection_type', 'collection_type'),
+    )
+
+    __mapper_args__ = {'polymorphic_on': 'collection_type'}
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+class DatasetMixin:
+    """
+        Mixin class for common Dataset attributes.
+    """
+
+    default_benchmark = Column(String, nullable=True)
+    default_keywords  = Column(JSON, nullable=True)
+
+    default_driver = Column(String, nullable=True)
+    default_units = Column(String, nullable=True)
+    alias_keywords = Column(JSON, nullable=True)
+    default_program  = Column(String, nullable=True)
+
+    contributed_values = Column(JSON)
+
+    history_keys = Column(ARRAY(String, as_tuple=True))
+    history = Column(ARRAY(String, as_tuple=True))
+
+
+class DatasetRecordsORM(Base):
+    """Association table for many to many"""
+
+    __tablename__ = 'dataset_records'
+
+    dataset_id = Column(Integer, ForeignKey('dataset.id', ondelete='cascade'), primary_key=True)
+    #TODO: check the cascase_delete with molecule
+    molecule_id = Column(Integer, ForeignKey('molecule.id', ondelete='cascade'), primary_key=True)
```

Comment:
On the fence here with cascade delete. I guess if someone makes a bad datasets, scraps it and then wants to build another one there are molecules that would not be useful to keep around. On the flip side if we build a dataset and run computations with it, cascade delete here will be stopped as the results will be linked to this molecule as well.

If a cascade delete fails are there error involved or is it silent in nature?

Reply:
At the same time this will be different than a reactiondataset where the molecules will not be cleaned due to their BSON structures?

Reply:
Ok, I'll change history and history_keys to JSON.
This cascade IN FOREIGN KEYS means: if my parent is deleted (Dataset or Molecule) then delete this DatasetRecord, not the way around. I'm sorry if this is confusing. 
This might be a rare case that we delete a molecule, but surely we might delete a dataset.
When we delete a dataset, it will delete it's records.

Cascade in RELATIONS means: delete those children when I'm (the parent) is deleted. It's different than cascade in foreign keys.

Reply:
Just to 100% confirm. If I try to delete a molecule that is in a DatasetRecord the molecule delete will be blocked?

Reply:
We can control this.
What's implemted now is: when a molecule in the molecule table is deleted, it will be automatically removed from the DatasetRecords.

We can change that behavior to make it though an error and we won't be able to delete a molecule that has a DatasetRecord. (by simply removing the cascade=delete)

Reply:
Ok gotcha, thats what I thought it said, I think we should remove `cascade=delete` here. So that people do not accidentally mess with their collections when removing molecules.